keps/sig-api-machinery/4153-declarative-validation/README.md

```diff
@@ -0,0 +1,1793 @@
+
+xref: [Kubernetes mapping of OpenAPI value validations to response code](https://github.com/kubernetes/kubernetes/blob/7adcb3cb37084750ab572a179ba01b9878b8de5f/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go#L60)
+
+### Unions
+
+Unions are informally defined but common in built-in types.
+
+We will convert most union validation to CEL expressions. For example:
+
+```go
+// staging/src/k8s.io/api/apps/v1/types.go
+
+type DeploymentSpec struct {
+  // ...
+  //
+  // +minimum=0
+  Replicas string `json...`
+  
+  // ...
+  // 
+  // +validationRule="!(self.type == 'Recreate') || !has(self.rollingUpdate)",message="may not be specified when strategy `type` is Recreate",reason=Forbidden,field="rollingUpdate"
+  // +validationRule="!(self.type == 'RollingUpdate') || has(self.rollingUpdate)",message="this should be defaulted and never be nil",reason=Required,field="rollingUpdate"
+  Strategy DeploymentStrategy `json:...`
+  
+  // ...
+}
+```
+
+For backward compatibility with existing validations, `reason` and `field` will
+be added to make it possible to declare a validation rule in CEL that matches
+not just the logic of the hand written validation it replaces, but also the
+exact field path and reason type.
+
+### CEL Validation
+
+CEL will play a major role in declarative validation by offering a way to
+declare validation rules for use cases that are too complex to be declared using
+JSON Schema value validations. Using CEL to validate APIs has been successfully
+demonstrated by CRD Validation Rules, which are on track for GA in 1.28.
+
+CEL rules will typically be placed on struct or field declarations and access
+multiple fields nested below the level of the type or field where the CEL
+expression is declared, e.g.:
+
+```go
+  
+  //+validationRule="!self.widgetType == 'Component' || !['badname2', 'badname2'].exists(notAllowed, self.componentName.contains(notAllowed))",reason=Forbidden
+  type FizzBuzzSpec struct {
+    Type WidgetType `json...`
+    ComponentName string `json...`
+  }
+```
+
+Note that the `+validationRule` IDL tags supports multiple optional fields after the expression string. The supported fields are:
+
+- `message`: A vailidation failure message string.
+- `messageExpression`: A CEL expression that evalutes to a validation failure message string. Takes precedence over message.
+- `reason`: Must be one of [Required, Forbidden, Invalid, RequestEntityTooLarge].
+- `field`: A JSONPath to the sub-field that a validation failure should be attributed to.
+
+We will need to extend our CEL libraries to make it possible to migrate all the
+validation rules that exist in the Kubernetes API today.
+
+- `isFormat() <bool>` and `validateFormat() <list<string>>` will be added to allow formats to be checked in CEL
+  expression and for format violations to be reported using `messageExpression: "self.validateFormat('ipv6')"`
+- IP and CIDR library:
+  - `isIP(string) bool`
```

Comment:
I think we will want to put some real thought into the patterns we want to use to do some of the more esoteric validations.  Like how would you express "must be a valid IP-string and must not be loopback" ?

`isIP(self) && !ip(string).isLoopback()`

That doesn't seem too bad, but what if I need to ensure a CIDR - the IP must be value, the range must be at least 4 bits but less than 16 if IPv4 and less than 64 if IPv6?

`isCDIR(self) && cidr(self).hostBits() >= 4 && ((cidr(self).is4() && cidr(self).hostBits() < 16 || cidr(self).is6() && cidr(self).hostBits() < 64))`

Yeesh.  Is there a `with` pattern in CEL?  Like

`with(cidr = cidr(self), with(hostBits = cidr.hostBits,  hostBits >= 4 && hostBits < lookup(map{4:16, 6:64}, hostBits)`

I'm just spitballing, but some of these expressions are going to be complicated.

Reply:
I agree with your point but I don't know if these are good formats. Not sure how they should be formulated. Helper methods to CEL rules?

Reply:
One precedent would be that ValidatingAdmissionPolicy has `variables` which allow you to predefine commonly shared expressions as a variable: https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/#variable-composition

Reply:
Yeah, I am mostly asking that we make it a goal not just to convert to CEL but to make the CEL readable and manageable, which might require finding "idioms" for complex relationships.

Reply:
Added a note and made this a top level goal to make the annotations actually nice to use.